tracer/contrib/redigo/tracedredis.go

```diff
@@ -0,0 +1,114 @@
+package tracedredis
+
+import (
+	"bytes"
+	"context"
+	"github.com/DataDog/dd-trace-go/tracer"
+	"github.com/DataDog/dd-trace-go/tracer/ext"
+	redis "github.com/garyburd/redigo/redis"
+	"net"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+// TracedConn is used to trace requests made to redis, it implements the interface redis.Conn.
+type TracedConn struct {
+	redis.Conn
+	traceParams TraceParams
+}
+
+// TraceParams contains the params useful for tracing.
+type TraceParams struct {
+	tracer  *tracer.Tracer
+	service string
+	network string
+	host    string
+	port    string
+}
+
+// TracedDial will return a TracedConn, it is meant to replace the redis.Dial function.
+func TracedDial(service string, tracer *tracer.Tracer, network, address string, options ...redis.DialOption) (TracedConn, error) {
+	c, err := redis.Dial(network, address)
+	addr := strings.Split(address, ":")
+	host := addr[0]
+	port := addr[1]
+	tracer.SetServiceInfo(service, "redis", ext.AppTypeDB)
+	tc := TracedConn{c, TraceParams{tracer, service, network, host, port}}
+	return tc, err
+}
+
+// SetService allows to change the spans service set in TracedDial on a TracedConn.
+func (tc *TracedConn) SetService(service string) {
+	tc.traceParams.tracer.SetServiceInfo(service, "redis", ext.AppTypeDB)
+	tc.traceParams.service = service
+}
+
+// TracedDialURL will return a TracedConn, this is the traced version of redis.DialURL.
+func TracedDialURL(service string, tracer *tracer.Tracer, rawurl string, options ...redis.DialOption) (TracedConn, error) {
+	u, err := url.Parse(rawurl)
+	if err != nil {
+		return TracedConn{}, err
+	}
+
+	// Getting host and port, usind code from https://github.com/garyburd/redigo/blob/master/redis/conn.go#L226
+	host, port, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		host = u.Host
+		port = "6379"
+	}
+	if host == "" {
+		host = "localhost"
+	}
+	// Set in redis.DialUrl source code
+	network := "tcp"
+	c, err := redis.DialURL(rawurl, options...)
+	tc := TracedConn{c, TraceParams{tracer, service, network, host, port}}
+	return tc, err
+}
+
+// Do overwrites redis.Do function and sends a span to the tracer.
+func (tc TracedConn) Do(commandName string, args ...interface{}) (reply interface{}, err error) {
+	var ctx context.Context
+	var ok bool
+	if len(args) > 0 {
+		ctx, ok = args[len(args)-1].(context.Context)
```

Comment:
I'm not familiar with that API but indeed having ctx as a last optional arg can lead to surprising behaviors. The reason for doing this is keeping the API unchanged when using or not using tracing, right ? Because this would work one way -> if you replace the `Conn` by a `TracedConn`, code is still going to work. But then all redis traces are lone traces, not related to their parents. So you have to put an extra arg. But if you do so, it does not work the other way, the day you go backwards and replace the `TracedConn` by a `Conn`, the code will fail because the Redis lib is going to receive an extra param it does not know what to do with. Dig into it, but probably `DoContext` suggested by @talwai would be fine, it would cost some extra typing/refactor from people instrumenting the code, but would make it safer. Just my 2c.